core/image/jpeg/jpeg.odin

```diff
@@ -0,0 +1,927 @@
+package jpeg
+
+import "core:bytes"
+import "core:compress"
+import "core:fmt"
+import "core:math"
+import "core:mem"
+import "core:image"
+import "core:slice"
+import "core:strings"
+
+Image :: image.Image
+Error :: image.Error
+Options :: image.Options
+
+HUFFMAN_MAX_SYMBOLS :: 162
+HUFFMAN_MAX_BITS  :: 16
+// 768 bytes of 24-bit RGB values.
+THUMBNAIL_PALETTE_SIZE :: 768
+BLOCK_SIZE :: 8
+COEFFICIENT_COUNT :: BLOCK_SIZE * BLOCK_SIZE
+
+Coefficient :: enum u8 {
+	DC,
+	AC,
+}
+
+Component :: enum u8 {
+	Y = 1,
+	Cb = 2,
+	Cr = 3,
+}
+
+HuffmanTable :: struct {
+	symbols: [HUFFMAN_MAX_SYMBOLS]byte,
+	codes: [HUFFMAN_MAX_SYMBOLS]u32,
+	offsets: [HUFFMAN_MAX_BITS + 1]byte,
+}
+
+QuantizationTable :: [COEFFICIENT_COUNT]u16be
+
+ColorComponent :: struct {
+	dc_table_idx: u8,
+	ac_table_idx: u8,
+	quantization_table_idx: u8,
+	v_sampling_factor: int,
+	h_sampling_factor: int,
+}
+
+Block :: [Component][COEFFICIENT_COUNT]i16
+
+@(private="file")
+zigzag := [?]byte{
+    0,   1,  8, 16,  9,  2,  3, 10,
+    17, 24, 32, 25, 18, 11,  4,  5,
+    12, 19, 26, 33, 40, 48, 41, 34,
+    27, 20, 13,  6,  7, 14, 21, 28,
+    35, 42, 49, 56, 57, 50, 43, 36,
+    29, 22, 15, 23, 30, 37, 44, 51,
+    58, 59, 52, 45, 38, 31, 39, 46,
+    53, 60, 61, 54, 47, 55, 62, 63,
+}
+
+@(optimization_mode="favor_size", private="file")
+refill_msb :: #force_inline proc(z: ^compress.Context_Memory_Input, width := i8(48)) {
+	refill := u64(width)
+	b      := u64(0)
+
+	if z.num_bits > refill {
+		return
+	}
+
+	for {
+		if len(z.input_data) != 0 {
+			b = u64(z.input_data[0])
+
+			if len(z.input_data) > 1 {
+				next := u64(z.input_data[1])
+
+				if b == 0xFF {
+					if next == 0x00 {
+						// 0x00 is used as a stuffing to indicate that the 0xFF is part of the data and not
+						// the beginning of a marker
+						z.input_data = z.input_data[2:]
+					} else if next >= cast(u64)image.JPEG_Marker.RST0 && next <= cast(u64)image.JPEG_Marker.RST7 {
+						// Skip any RSTn markers if we encounter them
+						b = u64(z.input_data[2])
+						z.input_data = z.input_data[3:]
```

Comment:
Malicious JPEG can hit an OOB here because of a missing check to ensure `len(z.input_data) > 2`